dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,417 @@
+  } else if (missing_component_behavior == "empty") {
+    return MissingExtBehavior::Empty;
+  } else if (missing_component_behavior == "error") {
+    return MissingExtBehavior::Raise;
+  } else {
+    return MissingExtBehavior::Invalid;
+  }
+}
+
+inline SampleConfig ParseSampleConfig(std::ifstream& config, const std::string& config_path) {
+  SampleConfig out;
+  DALI_ENFORCE(config >> out.start_offset,
+               make_string("Malformed index file at ", config_path,
+                           " - less components than stated at the beginning of the index file"));
+  out.end_offset = std::numeric_limits<decltype(out.end_offset)>::max();
+
+  std::string component_metadata;
+  std::getline(config, component_metadata);
+
+  std::stringstream extensions_stream(component_metadata);
+  std::string extension;
+  int64_t size;
+
+  while (extensions_stream >> extension) {
+    DALI_ENFORCE(extensions_stream >> size,
+                 make_string("Malformed index file at ", config_path,
+                             " - size corresponding to the extension '", extension, "' not found"));
+    out.config_metadata.push_back(ComponentConfig(std::move(extension), size));
+    size = 0;
+  }
+  DALI_ENFORCE(out.config_metadata.size() > 0_uz,
+               make_string("Malformed index file at ", config_path,
+                             " - no extensions provided for the sample"));
+  return out;
+}
+
+inline std::vector<SampleConfig> ParseConfig(MissingExtBehavior missing_component_behavior,
+                                             const std::string& config_path,
+                                             std::vector<std::set<std::string>>& ext) {
+  std::ifstream config(config_path);
+  int64_t config_size;
+  size_t sample_config_num = 0;  // for config validity check
+  config >> config_size >> sample_config_num;
+  DALI_ENFORCE(sample_config_num > 0_uz,
+               "Empty index file at " + config_path);  // config validity check
+
+  std::vector<SampleConfig> out;
+  out.reserve(sample_config_num);
+
+  for (size_t sample_index = 0; sample_index < sample_config_num; sample_index++) {
+    SampleConfig new_sample = ParseSampleConfig(config, config_path);
+    if (!out.empty()) {
+      out.back().end_offset = std::min(out.back().end_offset, new_sample.start_offset);
+    }
+
+    // filtering out samples without the required extensions
+    bool discard = false;
+    if (!std::all_of(ext.begin(), ext.end(), [&](std::set<std::string> extension_set) {
+          for (auto& component_config : new_sample.config_metadata) {
+            if (extension_set.count(component_config.ext)) {
+              return true;
+            }
+          }
+          return false;
+        })) {
+      switch (missing_component_behavior) {
+        case MissingExtBehavior::Skip:
+          discard = true;
+          break;
+        case MissingExtBehavior::Raise:
+          DALI_FAIL("Underful sample detected in index file at " + config_path);
```

Comment:
Maybe you can print line number as well?

Reply:
alright